File: packages/sdk/src/surfaces.ts
import type { Mindwire } from "./client.js";
import { readSSE } from "./sse.js";

export interface SurfaceGeometry { width: number; height: number; revision: number }
export interface SurfaceProblem { code: string; message: string }
export interface SurfaceCapabilities {
  view: boolean; capture: boolean; pointer: boolean; keyboard: boolean; text: boolean;
  clipboardRead: boolean; clipboardWrite: boolean;
}
export interface SurfaceController {
  sessionId: string; actor: "user" | "agent"; name: string; runId?: string; chatId?: string;
  generation: number; expiresAt: string;
}
export interface SurfaceSnapshot {
  id: string; workspaceId: string; kind: "desktop"; provider: "oblien";
  version: number; revision: number; instanceId: string;
  state: string; observedAt?: string; supported: boolean; enabled: boolean; available: boolean;
  credentials: boolean; os?: string; capabilities: SurfaceCapabilities;
  geometry?: SurfaceGeometry; controller?: SurfaceController;
  authorizationExpiresAt?: string; error?: SurfaceProblem;
}
/** Write-only, desktop-only, expiring provider grant. Never pass a user's session JWT. */
export interface SurfaceBinding {
  registryId: string; workspaceId: string; gatewayToken?: string;
  connection: {
    expires_at: string;
    ssh: { host: string; port: number; username: string; password: string; host_key_fingerprint: string };
    vnc: { host: string; port: number; authentication: "none" };
  };
}
export interface SurfaceSession {
  id: string; surfaceId: string; actor: "user" | "agent"; name: string;
  chatId?: string; runId?: string; mode: "view" | "control"; createdAt: string;
  controller?: SurfaceController;
}
export interface SurfaceOpenRequest { requestId: string; name?: string; mode: "view" | "control" }
export interface SurfaceControlRequest { action: "acquire" | "takeover" | "release" | "renew"; generation?: number }
export interface SurfaceAction {
  kind: "pointer" | "click" | "drag" | "scroll" | "key" | "text" | "clipboard_read" | "clipboard_write" | "release";
  x?: number; y?: number; toX?: number; toY?: number; buttons?: number;
  button?: "left" | "middle" | "right"; count?: number; deltaX?: number; deltaY?: number;
  keys?: string[]; text?: string;
}
export interface SurfaceActionRequest {
  /** Reuse this ID to inspect/recover an uncertain acknowledgement. Do not replay with a new ID. */
  requestId: string; sessionId: string; controlGeneration: number; frameId?: string;
  geometryRevision?: number; action: SurfaceAction;
}
export interface SurfaceReceipt {
  id: string; sessionId: string; surfaceId: string; kind: string;
  status: "dispatching" | "dispatched" | "outcome_unknown"; createdAt: string; error?: SurfaceProblem;
  /** Transient clipboard output, not stored in the receipt. */
  text?: string;
}
export interface SurfaceCapture {
  id: string; surfaceId: string; artifactId: string; mime: string; geometry: SurfaceGeometry; createdAt: string;
}
export interface ArtifactContent {
  id: string; chatId?: string; mime: string; bytes: number; width: number; height: number; createdAt: string;
  /** Base64 bytes, compatible with every existing daemon transport. */
  data: string;
}
export interface SurfaceToolAction {
  surfaceId: string; operation: string; sessionId?: string; receiptId?: string; status?: string;
  capture?: { id: string; artifactId: string; width: number; height: number };
}

/** Workspace-scoped API. One controller is shared by every harness and app client. */
export class SurfacesApi {
  constructor(private readonly mw: Mindwire) {}
  list(): Promise<SurfaceSnapshot[]> { return this.mw.http.request("GET", "/surfaces"); }
  status(refresh = false): Promise<SurfaceSnapshot> {
    return this.mw.http.request("GET", "/surfaces/desktop", { query: { refresh } });
  }
  bind(binding: SurfaceBinding): Promise<SurfaceSnapshot> {
    return this.mw.http.request("PUT", "/surfaces/desktop/binding", { body: binding });
  }
  open(request: SurfaceOpenRequest): Promise<SurfaceSession> {
    return this.mw.http.request("POST", "/surfaces/desktop/sessions", { body: request });
  }
  control(id: string, request: SurfaceControlRequest): Promise<SurfaceSession> {
    return this.mw.http.request("POST", `/surfaces/desktop/sessions/${encodeURIComponent(id)}/control`, { body: request });
  }
  close(id: string): Promise<{ closed: boolean }> {
    return this.mw.http.request("DELETE", `/surfaces/desktop/sessions/${encodeURIComponent(id)}`);
  }
  capture(id: string): Promise<SurfaceCapture> {
    return this.mw.http.request("POST", `/surfaces/desktop/sessions/${encodeURIComponent(id)}/captures`);
  }
  action(request: SurfaceActionRequest): Promise<SurfaceReceipt> {
    return this.mw.http.request("POST", "/surfaces/desktop/actions", { body: request });
  }
  receipt(id: string): Promise<SurfaceReceipt> {
    return this.mw.http.request("GET", `/surfaces/desktop/actions/${encodeURIComponent(id)}`);
  }
  artifact(id: string): Promise<ArtifactContent> {
    return this.mw.http.request("GET", `/artifacts/${encodeURIComponent(id)}`);
  }
  /** Every connection starts with the current snapshot, then revisions. Never replays input. */
  async *watch(opts: { signal?: AbortSignal } = {}): AsyncGenerator<SurfaceSnapshot> {
    const controller = new AbortController();
    const abort = () => controller.abort();
    opts.signal?.addEventListener("abort", abort, { once: true });
    if (opts.signal?.aborted) controller.abort();
    try {
      const response = await this.mw.http.open("GET", "/surfaces/desktop/events", { signal: controller.signal });
      let instance: string | undefined;
      let revision = -1;
      for await (const snapshot of readSSE<SurfaceSnapshot>(response.body!, controller.signal)) {
        if (instance !== snapshot.instanceId || snapshot.revision > revision) {
          instance = snapshot.instanceId; revision = snapshot.revision; yield snapshot;
        }
      }
    } finally {
      controller.abort(); opts.signal?.removeEventListener("abort", abort);
    }
  }
}

File: daemon/sdk/surfaces.go
package mindwire

import (
	"context"
	"github.com/oblien/mindwire/daemon/internal/artifact"
	"github.com/oblien/mindwire/daemon/internal/surface"
)

type SurfaceSnapshot = surface.Snapshot
type SurfaceSession = surface.Session
type SurfaceController = surface.Controller
type SurfaceGeometry = surface.Geometry
type SurfaceCapabilities = surface.Capabilities
type SurfaceBinding = surface.Binding
type DesktopSSHConnection = surface.SSHConnection
type DesktopSSHSettings = surface.SSHSettings
type DesktopVNCSettings = surface.VNCSettings
type SurfaceOpenRequest = surface.OpenRequest
type SurfaceControlRequest = surface.ControlRequest
type SurfaceAction = surface.Action
type SurfaceActionRequest = surface.ActionRequest
type SurfaceReceipt = surface.Receipt
type SurfaceCapture = surface.Capture
type SurfaceError = surface.Error
type ArtifactContent = artifact.Content

// Surfaces is workspace scoped. All WithAgent views share the same controller.
type Surfaces struct{ c *Client }

var surfaceUser = surface.Actor{Kind: "user", Name: "You"}

func (s *Surfaces) List() []SurfaceSnapshot { return []SurfaceSnapshot{s.c.core.surfaces.Snapshot()} }
func (s *Surfaces) Status(ctx context.Context, refresh bool) (SurfaceSnapshot, error) {
	if refresh {
		snapshot, _ := s.c.core.surfaces.Refresh(ctx)
		return snapshot, nil
	}
	return s.c.core.surfaces.Snapshot(), nil
}
func (s *Surfaces) Bind(ctx context.Context, binding SurfaceBinding) (SurfaceSnapshot, error) {
	if err := s.c.core.surfaces.Bind(binding, s.c.core.store); err != nil {
		return SurfaceSnapshot{}, surfaceAPIError("Surfaces.Bind", err)
	}
	snapshot, _ := s.c.core.surfaces.Refresh(ctx)
	return snapshot, nil
}
func (s *Surfaces) Open(ctx context.Context, req SurfaceOpenRequest) (SurfaceSession, error) {
	value, err := s.c.core.surfaces.Open(ctx, surfaceUser, req)
	return value, surfaceAPIError("Surfaces.Open", err)
}
func (s *Surfaces) Control(ctx context.Context, id string, req SurfaceControlRequest) (SurfaceSession, error) {
	value, err := s.c.core.surfaces.Control(ctx, id, surfaceUser, req)
	return value, surfaceAPIError("Surfaces.Control", err)
}
func (s *Surfaces) Close(id string) error {
	return surfaceAPIError("Surfaces.Close", s.c.core.surfaces.CloseSession(id, surfaceUser))
}
func (s *Surfaces) Capture(ctx context.Context, id string) (SurfaceCapture, error) {
	value, err := s.c.core.surfaces.Capture(ctx, id, surfaceUser)
	return value, surfaceAPIError("Surfaces.Capture", err)
}
func (s *Surfaces) Action(ctx context.Context, req SurfaceActionRequest) (SurfaceReceipt, error) {
	value, err := s.c.core.surfaces.Apply(ctx, surfaceUser, req)
	return value, surfaceAPIError("Surfaces.Action", err)
}
func (s *Surfaces) Receipt(id string) (SurfaceReceipt, error) {
	value, err := s.c.core.surfaces.Receipt(id)
	return value, surfaceAPIError("Surfaces.Receipt", err)
}
func (s *Surfaces) Artifact(id string) (ArtifactContent, error) {
	value, err := s.c.core.surfaces.Artifacts().Get(id)
	return value, surfaceAPIError("Surfaces.Artifact", err)
}
func (s *Surfaces) Changes() <-chan struct{} { return s.c.core.surfaces.Changes() }

func surfaceAPIError(op string, err error) error {
	if err == nil {
		return nil
	}
	detail, status := surface.ErrorResponse(err)
	return &APIError{Op: op, Status: status, Message: detail.Message, Cause: err}
}
